#' Download example data
#'
#' `load_exdata` downloads and saves example data as temporary files. The list of directories returned contain the files in order:
#'  \describe{
#'     \item{8_trunc.npz}{Numpy sparse matrix generated by `methSCAn` python package with dimensions 129300644x833. Contains CpG methylation of chromosome 8 from a murine adult NSC lineage. Information was reduced to the genomic region 8628100-8684100.}
#'     \item{8_GpC_trunc.npz}{Numpy sparse matrix generated by `methSCAn` python package with dimensions 1129301117x797. Contains GpC methylation of chromosome 8 from a murine adult NSC lineage. Information was reduced to the genomic region 8628100-8684100.}
#'     \item{cell-annotation.tsv}{Tsv file containing cell ids and pseudotime of the cells used for the scNMT-sequencing to receive the original Numpy sparse matrices.}
#' }
#'
#' @return List of directories, in which the temporary files are saved.
#' @export
#'
#' @source Original files received from \href{https://doi.org/10.1101/2022.07.13.499860}{(Kremer et al., 2022)}. The sparse matrices were simplified by removing the values outside the chosen region.
#'
#' @references Kremer, L. P., Cerrizuela, S., Shukairi, M. E. A., Ellinger, T., Straub, J., Dehler, S., Korkmaz, A., Weichenhan, D., Plass, C., Anders, S., & Martin-Villalba, A. (2022). <br>
#' Single-cell triple-omics uncovers DNA methylation as key feature of stemness in the healthy and ischemic adult brain. <br>
#' bioRxiv, 2022.2007.2013.499860 \url{https://doi.org/10.1101/2022.07.13.499860}.
#'
#'
#' @examples load_exdata()
load_exdata <- function() {

  url <- "https://heibox.uni-heidelberg.de/f/5f14053387914445b89f/?dl=1"
  tempfile <- file.path(tempdir(), "8_trunc.npz")
  download.file(url, tempfile, mode = "wb")


  url2 <- "https://heibox.uni-heidelberg.de/f/b76f7887a3f746b98e84/?dl=1"
  tempfile2 <- file.path(tempdir(), "8_GpC_trunc.npz")
  download.file(url2, tempfile2, mode = "wb")

  url3 <- "https://heibox.uni-heidelberg.de/f/fdc8bb53bf784cb092ab/?dl=1"
  tempfile3 <- file.path(tempdir(), "cell-annotation.tsv")
  download.file(url3, tempfile3, mode = "wb")

  # Return the loaded data
  return(list(tempfile, tempfile2, tempfile3))
}
